### ISSUE TITLE
One question about the experiments

### ISSUE THREAD

**Issue by typoverflow** [LanqingLi1993/FOCAL-ICLR#5]
I have a question about the experiments in the paper. It seems that during meta-testing the agent is provided with a pre-collected dataset as well, while in MBML or batch Pearl, the meta policy needs to collect data on its own during meta-testing. So I am wondering whether a pre-collected dataset is provided for MBML and batch Pearl as well, and if so, how do the algorithms make use of it?
Thank you for your time.

**Comment by LanqingLi1993** [comment]
Hi,

MBML and batch PEARL are all context-based meta-RL algorithms, which infer task identity in form of an embedding vector $z$ from context. Therefore, in principle, both algorithms can adapt to the fully-offline setting in FOCAL, by performing task inference using pre-collected dataset, without exploration. What you need to do is something similar to using the offline sampler defined in https://github.com/LanqingLi1993/FOCAL-ICLR/blob/master/rlkit/samplers/in_place.py.

As in section 5.1 of the FOCAL paper, we said
> The first is by naively modifying PEARL to train and test from logged data without exploration, which we term Batch PEARL

which means the batch PEARL in our experiments also uses pre-collected dataset at test time. As for MBML, I'm not 100% sure about which version we used. It's probably the fully-offline version as well, as MBML consistently underperformed in our experiments, compared to the result in the original paper (https://proceedings.neurips.cc/paper/2020/file/4496bf24afe7fab6f046bf4923da8de6-Paper.pdf).

If you are thinking of advancing the offline meta-RL benchmark, you might want to explore other competitive baselines as well, to name a few: FOCAL++[1], MACAW[2], BOReL[3], MerPO[4] and generalized decision transformer[5] . Good luck and let me know if you have further questions!  

[1] Li, Lanqing, et al. "Provably Improved Context-Based Offline Meta-RL with Attention and Contrastive Learning." arXiv e-prints (2021): arXiv-2102.
[2] Mitchell, Eric, et al. "Offline meta-reinforcement learning with advantage weighting." International Conference on Machine Learning. PMLR, 2021.
[3] Dorfman, Ron, Idan Shenfeld, and Aviv Tamar. "Offline meta learning of exploration." arXiv preprint arXiv:2008.02598 (2020).
[4] Lin, Sen, et al. "Model-Based Offline Meta-Reinforcement Learning with Regularization." arXiv preprint arXiv:2202.02929 (2022).
[5] Furuta, Hiroki, Yutaka Matsuo, and Shixiang Shane Gu. "Generalized Decision Transformer for Offline Hindsight Information Matching." arXiv preprint arXiv:2111.10364 (2021).

**Comment by typoverflow** [comment]
Thanks for your explanation, now I understand the meta-test procedure. 

However, concerning MBML, you said that
> As for MBML, I'm not 100% sure about which version we used. It's probably the fully-offline version as well, as MBML consistently underperformed in our experiments, compared to the result in the original paper.

This is somehow strange to me. As far as I know, MBML learns an encoder for task-inference from the pre-collected datasets, and I think the encoder should perform better when given data of similar quality (or from the same distribution). For example, in Sparse-Robot-Point where the reward is sparse, the encoder should make a better inference for the task's identification when given an expert dataset (which contains non-zero reward transitions), compared to when transitions are collected by the meta policy. So why is the performance inferior to what was reported in the original paper? 

I'm quite new to OMRL, please correct me if there is any misunderstanding : )

**Comment by LanqingLi1993** [comment]
Hi,

You are making a very good point about the Sparse-Point-Robot case. I agree that expert-level offline data with dense reward can be more informative for task inference than data collected online, but can be prone to overfitting as well [1]. However, if you go through the MBML paper, there is no experiment for Sparse-Point-Robot or other environments with sparse reward. My guess is that for OMRL algorithms on sparse-reward environments, meta-testing with offline or online data can be both challenging, in which FOCAL and probably MBML may exhibit suboptimal performance. 

Generalization in RL, especially in offline and meta-RL, is an open and challenging topic. We studied the generalization, namely the out-of-distribution and sparse reward problem of FOCAL in our follow-up paper [1], and I'm pretty sure there are quite a few emerging works worth attention. Also MBML is a relatively old baseline in the OMRL field. If you try to invent a new algorithm, consider comparing it with more concurrent algorithms which I listed in the last thread. Good luck!

[1] Li, Lanqing, et al. "Provably Improved Context-Based Offline Meta-RL with Attention and Contrastive Learning." arXiv e-prints (2021): arXiv-2102.
